Add a command to print a bunch of environment info

eslickj opened the pull request:
## Fixes issue #313

## Summary

Not sure I took the best approach here, so don't feel bad about making suggestions.

## Changes proposed in this PR:
Adds a command to print info about
-  IDAES version
- Python version
- OS version
- Pyomo version
- Dependency versions
- Extra package versions (by this I mean packages users have to install to use some things)

## Sample Output

```sh
(idaes-pse) C:\Users\EslickJ\git\idaes-pse\idaes\core\util>idaes environment-info

IDAES
    IDAES Version: 1.11.0.dev0
    IDAES Git Hash: 9ca7252853c51b04b9b2c265c73b581104e923fc

Pyomo
    Pyomo Version: 6.0.2.dev0

OS
    Platform: Windows
    Release: 10
    Version: 10.0.19041

Python
    Python Version: 3.9.4 (default, Apr  9 2021, 11:43:21) [MSC v.1916 64 bit (AMD64)]

Dependencies
    backports.shutil_get_terminal_size: 1.0.0
    bunch: 1.0.1
    click: 7.1.2
    colorama: 0.4.4
    flask: 2.0.0
    flask-cors: 3.0.10
    jupyter: 1.0.0
    lxml: 4.6.3
    matplotlib: 3.4.2
    nbconvert: 6.0.7
    nbformat: 5.1.3
    numpy: 1.20.3
    networkx: 2.5.1
    pandas: 1.2.4
    pint: 0.17
    psutil: 5.8.0
    pytest: 6.2.4
    pyyaml: 5.4.1
    requests: 2.25.1
    python-slugify: 5.0.2
    scipy: 1.6.3
    sympy: 1.8
    tinydb: 4.4.0
    rbfopt: 4.2.2
    pywin32: 225

Extras
    seaborn: Not Installed
```

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

dangunter commented:
@eslickj this looks great. I think, though, you should also update the docs in the `docs/user_guide/commands` directory

eslickj commented:
@dangunter Thanks!  Will add docs.

eslickj commented:
Ok, I updated the docs and added a section that check the solver versions of solvers in the binary distribution.  You can add more solvers to check with the "--solver" option.

lbianchi-lbl commented on idaes/core/util/env_info.py:
```diff
@@ -0,0 +1,130 @@
+        self.global_config = idaes._global_config_file
+        self.local_config = idaes._local_config_file
+        # Get Python info
+        self.python_version = sys.version
```
It could be useful to get the path to the Python executable (from `sys.executable`) here as well.

lbianchi-lbl commented on idaes/core/util/env_info.py:
```diff
@@ -0,0 +1,130 @@
+        "dot_sens",
+    ]
+
+    def __init__(self, addtional_solvers=()):
```
```suggestion
    def __init__(self, additional_solvers=()):
```

lbianchi-lbl commented on idaes/core/util/env_info.py:
```diff
@@ -0,0 +1,130 @@
+            except pkg_resources.DistributionNotFound:
+                self.extra_versions[dep] = None
+        self.solver_versions = {}
+        for s in self.known_solvers + list(addtional_solvers):
```
```suggestion
        for s in self.known_solvers + list(additional_solvers):
```

lbianchi-lbl commented on idaes/core/util/env_info.py:
```diff
@@ -0,0 +1,130 @@
+
+    def display_dict(self):
+        """Return a dictionary that is in a format that allows easy printing"""
+        d = OrderedDict([
```
Minor suggestions for streamlining code: these are completely optional, and I'm mostly adding them for information purposes.

- There's nothing wrong with using an `OrderedDict` to be more explicit, but I think a normal `dict` would be fine since `dict` is guaranteed to keep insertion order since Python 3.7 (this is effectively true also for Python 3.6,  although it's technically implementation-dependent)
- Given the structure of the data (a dict of dicts), defining `d = collections.defaultdict(dict)` (and optionally converting it to a regular dict at the end by having `return dict(d)`) should work and could make the code a bit simpler since you wouldn't have to define the empty keys manually

eslickj replied:
I didn't realized dictionaries were ordered now.  Thanks!  Streamlined things.

lbianchi-lbl commented on idaes/commands/env_info.py:
```diff
@@ -0,0 +1,50 @@
+)
+def environment_info(solver, file):
+    info = EnvironmentInfo(addtional_solvers=solver).display_dict()
+    if file is None:
+        for k, v in info.items():
+            click.echo("")
+            click.echo(f"{k}")
+            for l, q in v.items():
+                click.echo(f"    {l}: {q}")
+    else:
+        with open(file, "w") as f:
+            for k, v in info.items():
+                f.write("\n")
+                f.write(f"{k}\n")
+                for l, q in v.items():
+                    f.write(f"    {l}: {q}\n")
```
General suggestion: I think that having the tool return JSON output would be good, and could possibly come in handy in the future for e.g. ease of integration with CI tools, etc.

Since the data is already easily JSONable, adding an optional `--json` flag should be straightforward, and I'd even suggest considering making it the default since it's not too unreadable (especially with options such as  `indent=4`) compared to the pretty-printed output.

eslickj replied:
I switched the --file option to --json, which now writes the file in json format.

lbianchi-lbl commented on idaes/commands/env_info.py:
```diff
@@ -0,0 +1,50 @@
+    help="Write output ot a file"
+)
+def environment_info(solver, file):
+    info = EnvironmentInfo(addtional_solvers=solver).display_dict()
```
```suggestion
    info = EnvironmentInfo(additional_solvers=solver).display_dict()
```

lbianchi-lbl reviewed (changes requested):
I think this is a very useful tool that will be valuable when filing bugs or debugging in general both for the reporter and the maintainers.

I added changes on what I think is a typo in an argument name, plus other minor suggestions that I consider fully optional for of this PR.

lbianchi-lbl commented on idaes/core/util/env_info.py:
```diff
@@ -0,0 +1,145 @@
+                k: v if v is not None else "Not Installed"
+                for k, v in sorted(self.extra_versions.items())
+            },
+            "Sovlers": {
```
```suggestion
            "Solvers": {
```

lbianchi-lbl commented on idaes/core/util/env_info.py:
```diff
@@ -0,0 +1,145 @@
+            },
+            "Python": {
+                "Version": self.python_version,
+                "Executable": self.python_version,
```
```suggestion
                "Executable": self.python_executable,
```

lbianchi-lbl commented on idaes/core/util/env_info.py:
```diff
@@ -0,0 +1,145 @@
+# license information.
+#################################################################################
+
+__Author__ = "John Eslick"
```
```suggestion
__author__ = "John Eslick"
```

(it might be case-insensitive, but I'm not sure)

lbianchi-lbl reviewed (changes requested):
A couple of remaining typos, but other than that this looks good to go!

lbianchi-lbl reviewed (approved):
Looks good!

ksbeattie reviewed (approved):
LGTM
